Review of "Wrapping APIs" vignette (#70)
r-lib/httr2: vignettes/wrapping-apis.Rmd

  req_dry_run()
```

> JENNY COMMENT HERE

Author: Oh, no, I don't think you should change it.

I had gotten the impression there was *no* default user agent and one always had to set it. I think this just needs a sentence somewhere that, when you create an API-specific wrapper package, part of the infrastructure built into your main "request constructor" should include setting the user agent.

> JENNY COMMENT HERE

Reviewer: Ok, I'll just make it more clear that is in a fill in the blanks example.

### Core request function

Once you've made a few successful requests, it's worth seeing if you can figure out the general pattern so you can wrap it up into a function that will become the core of your package.

For faker, I spent a little with the [documentation](https://fakerapi.it/en#basic-usage) noting some commonalities:

-   Every URL is of the form [`https://fakerapi.it/api/v1`](https://fakerapi.it/api/v1)`/{resource}`, and data is passed to the resource with query parameters.
    All parameters start with `_`.

> -   Every resource has three common query parameters: `_locale`, `_quantity`, and `_seed` which have default values of `en_US`, `10`, and `null` respectively.

Author: So the default for `_quantity` is `10` not `1`? If so, that needs to change in a few other places.

Reviewer: The default for the API is 10; but I want to change it to 1 for these demos.

-   All endpoints return JSON data.

This lead me to construct the following function:

> JENNY COMMENT HERE

Reviewer: Ok, I've explained this a little more now.  Also explained what `NULL` does here.

There's one trick to notice here --- `req_url_query()` uses [dynamic dots](https://rlang.r-lib.org/reference/dyn-dots.html), so we can use `!!!` to convert (e.g.) `` req_url_query(req, !!!list(`_quantity` = 1, `_locale` = "en_US") `` into `` req_url_query(req, `_quantity` = 1, `_locale` = "en_US") ``.

> JENNY COMMENT HERE

Author: OK I guess was reading this as if you were demo'ing with `faker()`, which has this `_` prefixing business. But you actually just use `req_url_query()`, so the claim does make sense.

> JENNY COMMENT HERE

Reviewer: I've added an explanation of why I do it — here the motivation is primarily that it's a pain to type argument names that start with `_` since you have to backtick quote them.

But it's not very friendly for the package user who might not know anything about web APIs.
So typically the next step in the process is to wrap up some individual endpoints with their own functions.

JENNY COMMENT HERE

> For example, let's take the persons endpoint which has three additional parameters: `gender` (male or female), `birthday_start`, and `birthday_end`.

Author: Oh, right, we're letting The API randomly generate a birthday.

faker_person("male", quantity = 5)
```

The next steps would be to export and document this function; I'll leave that up to you.

## Secret management

We need to take a quick break from APIs to talk about secrets.

> Secrets are important, because every API (except for very simnple APIs like faker) is going to require that you identify yourself in some way typically with a username and password or an API key.

Author: ```suggestion
Secrets are important, because every API (except for very simnple APIs like faker) is going to require that you identify yourself in some way, typically with an API key or a token.
```

And even if you expect that your users will usually provide this information, you're still going to need your own credentials in order to actually test your package.

This system is probably overkill if you only have one secret that you need to share in one or two places.

> But you almost invariably accumulate more secrets over time, and more people and computers that you need to share it with, so I think spending little time to understand this system and set up it for your package will pay off in the long term.

Author: ```suggestion
But you almost invariably accumulate more secrets over time, and more people and computers that you need to share them with, so I think spending a little time to understand this system and set up it for your package will pay off in the long term.
```

### Basics

httr2 provides `secret_encrypt()` and `secret_decrypt()` to scramble secrets so that you can include them in your public source code without worrying that others can read them.
There are three basic steps to this process:

1.  You create an **encryption** key with `secret_make_key()` that is used to scramble and descramble secrets using symmetric cryptography:

    ```{r}
    key <- secret_make_key()
    key
    ```

    (Note that `secret_make_key()` uses a cryptographically secure random number generator provided by OpenSSL; it is not affected by R's RNG settings, and there's no way to make it reproducible.)

2.  You scramble your secrets with `secret_encrypt()` and store the resulting text directly in the source code of your package:

    ```{r}

>     secret <- secret_encrypt("this is a secret message", key)
>     secret

Author: ```suggestion
    secret_scrambled <- secret_encrypt("this is a secret message", key)
    secret_scrambled
```

I know this is a bit pedantic but I seriously think it helps when you're first absorbing all this.

    ```

3.  When needed, you descramble the the secret using `secret_decrypt()`:

    ```{r}

>     secret_decrypt(secret, key)

Author: ```suggestion
    secret_decrypt(secret_scrambled, key)
```

Author: It feels like a concrete example is needed around here, but maybe that gets taken care of soon enough. Specifically, the code above makes it look like you'll store the scrambled version in a variable, but IRL these will be string literals.

    ```

> ### Package secrets

Author: ```suggestion
### Package keys and secrets
```

> You can create an number of encryption keys, but I highly recommend that you create one key per package, which I'll call the **package** key.

Author: ```suggestion
You can create any number of encryption keys, but I highly recommend that you create one key per package, which I'll call the **package** key.
```

In this section, I'll show you how to store that key so that you (and your automated tests) can use it, but no one else can.

httr2 is built around the notion that this key should live in an environment variable.
So the first step is to make your package key available on your local development machine by adding a line to your your user-level `.Renviron` (which you can easily open with `usethis::edit_r_environ()`):

    YOURPACKAGE_KEY=key_you_generated_with_secret_make_key

```{r, include = FALSE}
Sys.setenv(YOURPACKAGE_KEY = secret_make_key())
```

> Now (after you restart R), you'll be able to take advantage of a special `secret_encrypt()` and `secret_decrypt()` feature: the encryption key can be the name of an environment variable.

Author: ```suggestion
Now (after you restart R), you'll be able to take advantage of a special `secret_encrypt()` and `secret_decrypt()` feature: the `key` argument can be the name of an environment variable, instead of the encryption key itself.
In fact, this is most natural usage.
```

Author: I'm not totally happy with "natural" but I want to convey "most intended", "most planned for", "should be most common".

```{r}

> secret <- secret_encrypt("my personal secret", "YOURPACKAGE_KEY")
> secret
> secret_decrypt(secret, "YOURPACKAGE_KEY")

Author: ```suggestion
secret_scrambled <- secret_encrypt("my personal secret", "YOURPACKAGE_KEY")
secret_scrambled
secret_decrypt(secret_scrambled, "YOURPACKAGE_KEY")
```

```

> You'll also need to make the key available in your GitHub Actions so your automated tests can use it.

Author: At any of these points, do we want to also point out the value of this whole system re: documentation? I.e. it make it clear you probably want to expose the encryption key to your pkgdown workflow, in addition to your check workflows.

This requires two steps:

1.  Add the key to your [repository secrets](https://docs.github.com/en/actions/reference/encrypted-secrets).

> 2.  Share the secret with the workflows that need it by adding a line to the appropriate workflow:

Author: ```suggestion
2.  Share the key with the workflows that need it by adding a line to the appropriate workflow:
```

    ``` yaml
        env:
          YOURPACKAGE_KEY: ${{ secrets.YOURPACKAGE_KEY }}
    ```

    You can see how httr2 does it in [its GitHub workflow](https://github.com/r-lib/httr2/blob/master/.github/workflows/R-CMD-check.yaml){.uri}.

> Other continuous integration platforms will offer similar ways to make secrets available in environment variables.

Author: ```suggestion
Other continuous integration platforms will offer similar ways to make secrets available in secure environment variables.
```

### When the package key isn't available

> There's an important computer that won't have your package key: CRAN.

Author: Is it also worth mentioning that this applies to PRs from external forks? If you don't do this, you will never get any decent info from CI for such PRs. Whereas with this, assuming you have some tests that don't require live API access, at least those tests will provide info and the others will be skipped vs. failing.

Author: Words I have used for this topic in https://gargle.r-lib.org/articles/articles/managing-tokens-securely.html

> Decryption is not going to work. Either because the Suggested sodium package is not available or (much more likely) because the environment variable that represents the key is not available.
> This will be the case on CRAN, by definition, because there is no way to share an encrypted secret.
> This will be the case for external contributors, on their personal machines and when their GitHub pull requests are checked via CI services, such as GitHub Actions, Travis-CI, or AppVeyor.
> ...
> By calling the skipper at the start of tests that require auth, you arrange for your package to cope gracefully when the token cannot be decrypted, e.g., on CRAN and in pull requests.

    secret_scrambled
    ```

3.  When needed, you descramble the the secret using `secret_decrypt()`:

    ```{r}
    secret_decrypt(secret_scrambled, key)
    ```

### Package keys and secrets

You can create any number of encryption keys, but I highly recommend that you create one key per package, which I'll call the **package** key.
In this section, I'll show you how to store that key so that you (and your automated tests) can use it, but no one else can.

httr2 is built around the notion that this key should live in an environment variable.
So the first step is to make your package key available on your local development machine by adding a line to your your user-level `.Renviron` (which you can easily open with `usethis::edit_r_environ()`):

    YOURPACKAGE_KEY=key_you_generated_with_secret_make_key

```{r, include = FALSE}
Sys.setenv(YOURPACKAGE_KEY = secret_make_key())
```

Now (after you restart R), you'll be able to take advantage of a special `secret_encrypt()` and `secret_decrypt()` feature: the `key` argument can be the name of an environment variable, instead of the encryption key itself.

> In fact, this is most natural usage.

Author: ```suggestion
In fact, this is the most natural usage.
```

Gah, another PR to a PR I've made to my own PR.

    ``` yaml
        env:
          YOURPACKAGE_KEY: ${{ secrets.YOURPACKAGE_KEY }}
    ```

    You can see how httr2 does it in <https://github.com/r-lib/httr2/blob/master/.github/workflows/R-CMD-check.yaml>.

Now you can encrypt your API key and store it in your tests --- there's no way for anyone to decrypt it unless they have your key.

JENNY COMMENT HERE

```{r}

> x_enc <- secret_encrypt("API-key-from-NYTimes-Books", "HTTR2_KEY")

Reviewer: Yup

When `secret_decrypt()` is run in tests, it will automatically skip the test if the key isn't found.
That means that your auth-requiring tests will automatically be skipped on CRAN.

### Security considerations

Note that including an API key as a query parameter is relatively insecure; if an API uses this method of auth, it's typically because the key is relatively easy to create or gives relatively few privileges.
Here it only takes a couple of minutes to generate your own NYTimes API key, so there's little incentive for someone to try and steal yours.

The main problem of conveying credentials via the url is that it's easily exposed, because httr2 makes no efforts to redact confidential information stored in query parameters.
This means it's relatively easy to leak your key if you use `req_perform(verbose = 1)`, `req_dry_run()`, or even just print the request object.

All in all, this means you need to be a little cautious if you need to get help with a reprex; you'll need to manually redact the key.

### Basic request

Now let's perform a test request and look at the response:

> JENNY COMMENT HERE

Reviewer: Ooops! I'll have to think about how to avoid this? Maybe some `req_redact()` function that's automatically applied to the output of `req_dryrun()`? 

Or maybe I should just embrace it, with a comment.

> JENNY COMMENT HERE

Reviewer: Decided to embrace it

To see if there's any extra useful information we can again suppress the default error and inspect the result:

JENNY COMMENT HERE

> 

Reviewer: I figured it out — I should be using `last_request()` as in the previous example.

> 

Author: That sounds right!

> You should sleep 6 seconds between calls to avoid hitting the per minute rate limit.
> If you need a higher rate limit, please contact us at [code\@nytimes.com](mailto:code@nytimes.com){.email}.

> JENNY SARCASTIC AND BITTER COMMENT HERE

Reviewer: I think it's getting more common, although even GitHub doesn't do it exactly according to the spec.

Putting together all the pieces above yields a function something like this:

```{r}
nytimes_books <- function(api_key, path, ...) {
  request("https://api.nytimes.com/svc/books/v3") %>% 
    req_url_path_append("/reviews.json") %>% 
    req_url_query(..., `api-key` = api_key) %>% 

>     # JENNY COMMENT HERE

Author: I see you do model this in a later example, but I wonder if you want to predefine the error-body-handling function? Either outside of `nytimes_books()` or inside. Because I can't imagine these would ever stay simple enough to define inline like this. Also this feels pretty hostile to writing tests.

Reviewer: Good call. I do that later on, but not here.

To finish this up for a real package, you'd want to:

-   Add explicit arguments and check that they have the correct type.

-   Export and document the function.

-   Convert the nested list into a more user-friendly data structure (probably a data frame with one row per review).

You'd also want to provide some convenient way for the user to supply their own API key.

### User-supplied key

A good place to start is an environment variable, because environment variables are easy to set without typing anything in the console (which can get accidentally shared via your `.Rhistory`) and are easily set in automated processes.
Then you'd write a function to retrieve the API key, returning a helpful message if it's not found:

JENNY COMMENT HERE

> 

Author: Can't remember what I wanted to say here.

Then you could modify `nytimes_books()` to use `get_api_key()` as the default value for `api_key`.
Since the argument is now optional, we can move it to end of the argument list, since it'll only be needed in exceptional circumstances.

> JENNY COMMENT HERE

Author: Maybe also worth pointing out "to the end and *after `...`*". If indeed that is intentional, but I suspect it is.

Reviewer: I was just using `...` casually here, but yes, I would recommend putting it after `...` if it exists (i.e. https://design.tidyverse.org/dots-position.html). Do you think I need to explicitly call that out here?

Author: Well, I really like this principle, so I'd take the chance to publicize it. But I also don't think matters much here. It's tangential.

  key <- Sys.getenv("NYTIMES_KEY")
  if (identical(key, "")) {
    if (is_testing()) {

>       # JENNY COMMENT HERE

Author: From the code I'm seeing here, it's not obvious to me that this condition (`is_testing()`) is sufficient to guarantee that `testing_key()` will successfully return a key. Have I missed something?

Reviewer: Nope! I forgot a `secret_has_key("HTTR2_KEY")`.

Reviewer: Ooops, no I didn't — `secret_decrypt()` will automatically skip if `HTTR2_KEY` is not found.

You have one username + password per site; you can have one token per use case.
This lets each use case have a minimal set of permissions, and you can easily revoke one token without affecting any other use case.

> JENNY COMMENT HERE

Author: Can't remember what I wanted to say here.

That body looks pretty useful!
So let's write a function that extracts the data and formats it for presentation to the user:

> JENNY COMMENT HERE

Author: It feels like all these functions should probably be prefixed with `gist_` instead of `github_`.

Also, kind of like rate limiting, I have yet to work with an API that structures its errors consistently enough to really do this. That would of course be great! I wonder how typical this 🤔 Of course, I know I will be able to take my extremely ugly version of this for Google errors and insert them into this framework. So what I'm really wondering is what these error-body-processing functions will really look like in the wild.

Reviewer: Agreed.

Yeah, and in my experience typically there's no way to find out how each endpoint actually structures its errors without deliberately forcing them. I'll add a note to that effect.

Author: And, with most APIs, it can be very hard to intentionally cause errors and capture them for study. You basically have to wait for yourself or a user to experience it and then respond.

### Sending data

To [create a gist](https://docs.github.com/en/rest/reference/gists#create-a-gist) we need to send a `POST` request with a body that contains data encoded as JSON.
httr2 provides one function that does both of these things: `req_body_json()`:

> JENNY COMMENT HERE

Author: At first, I understood "does both of those things" to suggest that `req_body_json()` does both of these things:
* **send** a `POST` request
* with a body ...

and I was surprised that it **sent** a request. But I was just confused. It only **forms** a `POST` request. Maybe this could be reworded.

## OAuth

> JENNY PEDANTIC COMMENT HERE

Author: I know the whole authentication vs. authorization thing and that we're sort of talking about both. But if we're going to pick one of those words to feature in this paragraph, here authorization somehow feels like the better choice?

Reviewer: Good point!

### Clients

The first step in working with any OAuth API is to create a client.
This involves you registering for a developer account on the API's website and creating a new OAuth app.
The process varies from API to API, but at the end of it you'll get a client id and in most cases a client secret.

> JENNY COMMENT HERE

Author: I feel like the scenario of "user provides own client" is not on the radar here at all, but it should be.

token <- oauth_flow_auth_code(client, auth_url = "https://github.com/login/oauth/authorize")
```

This flow can't be used inside a vignette because it's designed specifically for interactive use: it will open a webpage on GitHub that requires you to interactively confirm it's OK for this app to use your GitHub account.

Other flows provide different ways of getting the token:

> -   `req_oauth_client_credentials()` is used to allow the client to perform actions on its own behalf (instead of on behalf of some other user). JENNY COMMENT HERE

Author: I perceive the term "service account (token)" to be common enough to mention here. That's what I think this is about, yeah?

Reviewer: You mean explicitly connect this function to "service" accounts" or explain more what a service account is?

Author: Yeah, I just mean

> explicitly connect this function to "service" accounts"

-   The token includes an expiry date that's in the past.

-   The request returns a 401 status code.

> JENNY COMMENT HERE

Author: There seems to be some garbled / truncated wording in the bullet list above and the following 2 sentences.

If the access token has expired, httr2 will automatically retry automatically refresh before the first request.
If the request errors with a 401 and `invalid_token` error in the `WWW-authenticate` header.

### Caching

By default, `req_oauth_auth_code()` and friends will cache the token in memory, so that multiple requests in the same session all use the same token.
In some cases, you may want to save the token so that it's automatically used across sessions.
This is easy to do (just set `cache_disk = TRUE` in `req_oauth_auth_code()`) but you need to carefully consider the consequences of saving the user's credentials on disk.

httr2 does the best it can to save these credentials securely.
They are stored in a local cache directory (`rappdirs::user_cache_dir("httr2"))` that should only be accessible to the current user, and are encrypted so they will be hard for any package other than httr2 to read.

> However, there's no way to protect other R code from using httr2 to access them, so if you do choose to cache tokens, you should inform the user and give them the ability to opt-out.

Author: ```suggestion
However, there's no way to prevent other R code from using httr2 to access them, so if you do choose to cache tokens, you should inform the user and give them the ability to opt-out.
```

I saw this mis-wording once I was reviewing on GitHub.
